ls1intum/Artemis
# `Development`: Add support for evaluating BPMN diagrams in modeling exercises

This PR adds a parser to support BPMN diagrams modelling exercises and bumps the integrated Apollon version to 3.3.4.

### Checklist
#### General
#### Server
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).(https://docs.artemis.cit.tum.de/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).
#### Client

### Motivation and Context


To be able to use the full feature set offered from Artemis including Compass grading suggestions, the newly introduced BPMN diagram type also requires a corresponding parser.

### Description
This change adds a parser for BPMN diagram created with Apollon 3.x.x. No corresponding 2.x.x diagram schema parser has been implemented as no BPMN diagrams created with schema version 2.x.x exist.

### Steps for Testing

Prerequisites:
- 1 Student
- 1 Modelling Exercise with a BPMN diagram type

1. Log in to Artemis
2. Navigate to the course "BPMN Test" 
3. Navigate to the exercise "BPMN Modelling" (https://artemis-test6.artemis.cit.tum.de/courses/28/exercises/145)
4. Make a modelling submission
5. Ensure that the submission is accepted properly

### Testserver States
> [!NOTE]
> These badges show the state of the test servers.
> Green = Currently available, Red = Currently locked

![](https://byob.yarr.is/ls1intum/Artemis/artemis-test1)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test2)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test3)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test4)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test5)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test6)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test9)

### Review Progress






#### Performance Review
#### Code Review
#### Manual Tests
#### Exam Mode Test

### Test Coverage






| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|-----------------------------:|
| BPMNTransaction | 89% | ✅              |
| BPMNAnnotation | 89% | ✅                           |
| BPMNSubprocess | 89% | ✅              |
| BPMNDataObject | 89% | ✅              |
| BPMNCallActivity | 89% | ✅              |
| BPMNGroup | 89% | ✅              |
| BPMNSwimlane | 89% | ✅              |
| BPMNTask | 94% | ✅              |
| BPMNEndEvent | 94% | ✅              |
| BPMNGateway | 91% | ✅              |
| BPMNIntermediateEvent | 96% | ✅              |
| BPMNStartEvent | 91% | ✅              |
| BPMNPool | 94% | ✅              |
| BPMNFlow| 66% | ✅                           |
| BPMNDataStore | 94% | ✅              |
| BPMNIntermediateEventType | 97% | ✅              |
| BPMNTaskType | 96% | ✅              |
| BPMNEndEventType | 96% | ✅              |
| BPMNStartEventType | 95% | ✅              |
| BPMNGatewayType | 95% | ✅              |
| BPMNMarker | 94% | ✅              |
| BPMNFlowType | 94% | ✅              |
| BPMNDiagram | 98% | ✅              |
| BPMNDiagramParser | 96% | ✅              |
| **Total** | 92% |               |
